<?xml version="1.0" encoding="utf-8"?>
<launch>
    <arg name="veh" default="$(env VEHICLE_NAME)" />
    <arg name="config" default="baseline" />
    <arg name="param_file_name" default="default" />
    <arg name="verbose" default="false" />
    
    <!-- Adaptive Speed Controller with Safety Integration -->
    <group ns="$(arg veh)">
        
        <!-- Safety System Components (if not already running) -->
        <include file="$(find safety_system)/launch/safety_system.launch">
            <arg name="veh" value="$(arg veh)" />
            <arg name="config" value="$(arg config)" />
        </include>
        
        <!-- Predictive Perception System (if not already running) -->
        <include file="$(find predictive_perception)/launch/enhanced_perception_system.launch">
            <arg name="veh" value="$(arg veh)" />
            <arg name="config" value="$(arg config)" />
        </include>
        
        <!-- Adaptive Speed Controller Node -->
        <node name="adaptive_speed_controller_node" 
              pkg="adaptive_speed_control" 
              type="adaptive_speed_controller_node.py" 
              output="screen" 
              clear_params="true"
              required="true">
            
            <!-- Load configuration parameters -->
            <rosparam file="$(find adaptive_speed_control)/config/adaptive_speed_controller_node/$(arg param_file_name).yaml" command="load" />
            
            <!-- Topic remappings for safety integration -->
            <!-- Input topics -->
            <remap from="~lane_pose" to="/$(arg veh)/lane_filter_node/lane_pose" />
            <remap from="~camera_image" to="/$(arg veh)/camera_node/image/compressed" />
            <remap from="~vehicle_detections" to="/$(arg veh)/vehicle_detection_node/centers" />
            <remap from="~apriltag_detections" to="/$(arg veh)/apriltag_detector_node/detections" />
            <remap from="~stop_line_reading" to="/$(arg veh)/stop_line_filter_node/stop_line_reading" />
            <remap from="~imu_data" to="/$(arg veh)/imu_node/data" />
            
            <!-- Safety system integration -->
            <remap from="~collision_risk" to="/$(arg veh)/collision_detection_manager_node/collision_risk" />
            <remap from="~safety_status" to="/$(arg veh)/safety_fusion_manager_node/safety_status" />
            <remap from="~base_speed_command" to="/$(arg veh)/lane_controller_node/car_cmd" />
            
            <!-- Enhanced perception integration -->
            <remap from="~predicted_trajectories" to="/$(arg veh)/predictive_perception_manager_node/predicted_trajectories" />
            <remap from="~scene_analysis" to="/$(arg veh)/scene_understanding_module_node/scene_analysis" />
            
            <!-- Output topics -->
            <remap from="~adaptive_speed_command" to="/$(arg veh)/adaptive_speed_controller_node/car_cmd" />
            
            <!-- Parameters -->
            <param name="~verbose" value="$(arg verbose)" />
            <param name="~safety_integration_enabled" value="true" />
            <param name="~enable_environmental_adaptation" value="true" />
            <param name="~enable_following_distance" value="true" />
            <param name="~enable_smooth_acceleration" value="true" />
            
        </node>
        
        <!-- Safety Command Arbiter Integration -->
        <node name="safety_command_arbiter_node" 
              pkg="safety_system" 
              type="safety_command_arbiter_node.py" 
              output="screen"
              required="true">
            
            <!-- Input command sources -->
            <remap from="~lane_controller_cmd" to="/$(arg veh)/lane_controller_node/car_cmd" />
            <remap from="~navigation_cmd" to="/$(arg veh)/adaptive_speed_controller_node/car_cmd" />
            <remap from="~safety_override" to="/$(arg veh)/emergency_stop_system_node/safety_override" />
            
            <!-- Safety status inputs -->
            <remap from="~safety_status" to="/$(arg veh)/safety_fusion_manager_node/safety_status" />
            <remap from="~collision_risk" to="/$(arg veh)/collision_detection_manager_node/collision_risk" />
            
            <!-- Final output command -->
            <remap from="~car_cmd" to="/$(arg veh)/wheels_driver_node/wheels_cmd" />
            
            <!-- Configuration -->
            <param name="~max_linear_velocity" value="0.6" />
            <param name="~max_angular_velocity" value="2.0" />
            <param name="~safety_override_priority" value="10" />
            <param name="~enable_safety_limits" value="true" />
            
        </node>
        
        <!-- Performance Monitor for Safety Integration -->
        <node name="adaptive_speed_performance_monitor" 
              pkg="adaptive_speed_control" 
              type="performance_monitor_node.py" 
              output="screen"
              required="false">
            
            <!-- Monitor topics -->
            <remap from="~speed_commands" to="/$(arg veh)/adaptive_speed_controller_node/car_cmd" />
            <remap from="~safety_overrides" to="/$(arg veh)/safety_command_arbiter_node/safety_violations" />
            <remap from="~performance_stats" to="/$(arg veh)/adaptive_speed_controller_node/performance_stats" />
            
            <!-- Output monitoring data -->
            <remap from="~performance_report" to="/$(arg veh)/adaptive_speed_performance_monitor/report" />
            
        </node>
        
    </group>
</launch>